rust/connlib/tunnel/src/dns.rs

```diff
@@ -97,76 +78,126 @@ impl<T, V> ResolveStrategy<T, DnsQueryParams, V> {
+            ip_provider: IpProvider::for_resources(),
+        }
     }
 
-    let question = message.first_question()?;
-
-    tracing::trace!("Parsed packet as DNS query: '{question}'");
+    pub(crate) fn get_description(
+        &self,
+        ip: &IpAddr,
+        dns_resources: &HashMap<String, ResourceDescriptionDns>,
+    ) -> Option<ResourceDescriptionDns> {
+        let name = self.ips_to_fqdn.get(ip)?;
+        get_description(name, dns_resources)
+    }
+
+    pub(crate) fn get_fqdn(&self, ip: &IpAddr) -> Option<(&DomainName, &Vec<IpAddr>)> {
+        let fqdn = self.ips_to_fqdn.get(ip)?;
+        Some((fqdn, self.fqdn_to_ips.get(fqdn).unwrap()))
+    }
+
+    fn get_or_assign_ips(&mut self, fqdn: DomainName) -> Vec<IpAddr> {
+        let ips = self
+            .fqdn_to_ips
+            .entry(fqdn.clone())
+            .or_insert({
+                let mut ips = self.ip_provider.get_n_ipv4(4);
+                ips.extend_from_slice(&self.ip_provider.get_n_ipv6(4));
+                ips
+            })
+            .clone();
+        for ip in &ips {
+            self.ips_to_fqdn.insert(*ip, fqdn.clone());
+        }
 
-    // In general we prefer to always have a response NxDomain to deal with with domains we don't expect
-    // For systems with splitdns, in theory, we should only see Ptr queries we don't handle(e.g. apple's dns-sd)
-    let resource =
-        match resource_from_question(dns_resources, dns_resources_internal_ips, &question) {
-            Some(ResolveStrategy::LocalResponse(resource)) => Some(resource),
-            Some(ResolveStrategy::ForwardQuery(params)) => {
-                return Some(ResolveStrategy::ForwardQuery(params.into_query(packet)));
-            }
-            Some(ResolveStrategy::DeferredResponse(resource)) => {
-                return Some(ResolveStrategy::DeferredResponse((
-                    resource,
-                    question.qtype(),
-                )))
-            }
-            None => None,
-        };
-    let response = build_dns_with_answer(message, question.qname(), &resource)?;
-    Some(ResolveStrategy::LocalResponse(build_response(
-        packet, response,
-    )?))
-}
+        ips
+    }
 
-pub(crate) fn create_local_answer<'a>(
-    ips: &HashSet<IpAddr>,
-    packet: IpPacket<'a>,
-) -> Option<IpPacket<'a>> {
-    let datagram = packet.as_udp().unwrap();
-    let message = as_dns(&datagram).unwrap();
-    let question = message.first_question().unwrap();
-    let qtype = question.qtype();
-
-    #[allow(clippy::wildcard_enum_match_arm)]
-    let resource = match qtype {
-        Rtype::A => RecordData::A(
-            ips.iter()
+    // This function will panic if it's called with an invalid PTR question
+    fn get_records<N: ToName>(&mut self, question: &Question<N>) -> Vec<RecordData<DomainName>> {
+        match question.qtype() {
+            Rtype::A => self
+                .get_or_assign_ips(question.qname().to_name())
```

Comment:
As discussed. I think we should lazily allocate IPs and not assign IPv6 IPs if we are only asked for IPv4.

Reply:
I have to refactor how peer selection/connection work, will rework that then